### ISSUE TITLE
Type guard is not working properly with walrus conditional

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#686]
In issue #474, @alexkoay reported the following bug:

```
import re
def foo(s: str) -> str:
	if m := re.fullmatch('(test).+', s):
		return m.group(1)
	return 'oops'
```
group reports a reportOptionalMemberAccess error under Pyright 1.1.38.

**Comment by erictraut** [comment]
Looking into this more, the behavior is correct. The bug is in the code, not in Pyright. It was not reported as an error in previous versions of Pyright, but a recent bug fix in 1.1.38 changed that.

It has nothing to do with type guards or the walrus operator. The same error is identified in this simplified version of the code:
```
def foo(s: str):
    m = re.fullmatch('(test).+', s)
    m.group(1)
```

The problem here is that `re.fullmatch` can return `None`. If it does, your program will crash when it executes `m.group(1)`. Your code needs to deal with the None case. If you're confident that your usage of `re.fullmatch` will never produce a `None` result, you can add an assertion to the code, like this:
```
def foo(s: str):
    m = re.fullmatch('(test).+', s)
    assert m is not None
    m.group(1)
```

**Comment by decorator-factory** [comment]
These two are not equivalent at all!
```py
import re
def foo(s: str) -> str:
	if m := re.fullmatch('(test).+', s):
		return m.group(1)
	return 'oops'
```
```py
def foo(s: str):
    m = re.fullmatch('(test).+', s)
    m.group(1)
```
The first one will never execute `return m.group(1)`. 
I've had the same issue with:
```py
        if rematch := re.match(rf'^\s*{escaped}', string):
            _, end = rematch.span() # <-- "span" is not a known member of "None"
```
and even
```py
        if (rematch := re.match(rf'^\s*{escaped}', string)) is not None:
            _, end = rematch.span() # <-- "span" is not a known member of "None"
```

**Comment by erictraut** [comment]
You are correct. These aren't equivalent. Thanks for pointing that out. Re-opening the issue.

**Comment by decorator-factory** [comment]
Thank you!

**Comment by erictraut** [comment]
Type guards were already working for some assignment expressions, but it depended on the type of expression on the right-hand side. I've eliminated the restriction and added your example to the test suite.

The fix will be in the next version of Pyright.

**Comment by erictraut** [comment]
This is now fixed in Pyright 1.1.40, which I just published.
